File: scripts/notificar_jefe_evaluacion.py
"""
Envía correo al jefe directo de cada colaborador cuya primera y/o segunda
evaluación de desempeño ya llegó (fecha <= hoy).

Solo se envía si ya existen los documentos generados de "Evaluación de Desempeño"
y "Autoevaluación de Desempeño" para el colaborador. Si faltan, se reintenta en
las siguientes corridas diarias hasta que existan.

Solo se consideran fechas de evaluación ya llegadas (fecha <= hoy) y con un
atraso de hasta VENTANA_NOTIFICACION_DIAS (30) días — esto evita notificar a
colaboradores antiguos cuyas evaluaciones quedaron en el pasado y que ya se
gestionaron por otros medios.

Cada colaborador se notifica solo una vez por ciclo (1ra / 2da evaluación) al
jefe directo. También inserta notificaciones in-app en notificaciones_app.
"""
import os
import sys
import logging
from datetime import date, datetime
from html import escape

sys.path.insert(0, os.path.dirname(os.path.dirname(__file__)))
from config.db import get_connection
from scripts.mailer import send_email_smtp
from scripts.notificaciones_db import insert_notificaciones

LOG_FILE = os.path.join(os.path.dirname(os.path.dirname(__file__)), 'logs', 'notificacion_jefe_evaluacion.log')
logging.basicConfig(
    level=logging.INFO, format='%(asctime)s | %(message)s', datefmt='%Y-%m-%d %H:%M:%S',
    handlers=[logging.FileHandler(LOG_FILE, encoding='utf-8'), logging.StreamHandler()],
)
log = logging.getLogger(__name__)


GENERATED_DOCS_ROOT_CANDIDATES = [
    os.getenv('GENERATED_DOCS_ROOT', '').strip(),
    os.getenv('GENERATED_DOCS_DIR', '').strip(),
    os.path.normpath(os.path.join(os.path.dirname(__file__), '..', '..', 'talent-management-app', 'backend', 'generated_documents')),
    '/app/media/generated_documents',
    os.path.join(os.path.dirname(os.path.dirname(__file__)), 'generated_documents'),
]

CICLOS = [
    ('PRIMERA_EVALUACION', 'eval1_raw', 'Primera evaluación', '1ª Evaluación'),
    ('SEGUNDA_EVALUACION', 'eval2_raw', 'Segunda evaluación', '2ª Evaluación'),
]

# Solo se notifica si la fecha de la evaluación ya llegó y no tiene más de
# esta cantidad de días de atraso. Evita notificar evaluaciones de
# colaboradores antiguos cuyas fechas quedaron muy en el pasado y que ya
# se gestionaron por otros medios en su momento.
VENTANA_NOTIFICACION_DIAS = 30


def _resolve_doc_path(relative_path: str) -> str | None:
    if not relative_path:
        return None
    rel = relative_path.strip().replace('\\', os.sep).replace('/', os.sep)
    for root in GENERATED_DOCS_ROOT_CANDIDATES:
        if not root:
            continue
        full = os.path.join(root, rel)
        if os.path.isfile(full):
            return full
    return None


def _find_document(conn, employee_id: int, name_like: str, exclude_like: str | None = None) -> dict | None:
    sql = """
        SELECT TOP 1 relative_path, file_name
        FROM employee_generated_documents
        WHERE employee_id = ?
          AND template_name COLLATE Latin1_General_CI_AI LIKE ?
    """
    params = [employee_id, name_like]
    if exclude_like:
        sql += " AND template_name COLLATE Latin1_General_CI_AI NOT LIKE ?"
        params.append(exclude_like)
    sql += " ORDER BY updated_at DESC, created_at DESC"

    cursor = conn.cursor()
    cursor.execute(sql, params)
    row = cursor.fetchone()
    cursor.close()
    if not row:
        return None
    rel_path, file_name = row
    full_path = _resolve_doc_path(str(rel_path or ''))
    if not full_path:
        return None
    return {'path': full_path, 'name': str(file_name or os.path.basename(full_path))}


def _find_evaluation_documents(conn, employee_id: int) -> dict | None:
    """Retorna {'evaluacion': {...}, 'autoevaluacion': {...}} solo si AMBOS documentos
    existen para el colaborador; de lo contrario None."""
    autoeval = _find_document(conn, employee_id, '%autoevaluacion%')
    evaluacion = _find_document(conn, employee_id, '%evaluacion%desempeno%', exclude_like='%autoevaluacion%')
    if not autoeval or not evaluacion:
        return None
    return {'evaluacion': evaluacion, 'autoevaluacion': autoeval}


def _already_notified_once(conn, employee_id: int, task_code: str) -> bool:
    sql = """
        SELECT TOP 1 1
                FROM tareas_gestion_talento t
                WHERE t.task_code = ?
                    AND t.employee_id = ?
                    AND (
                                JSON_VALUE(COALESCE(t.metadata_json, '{}'), '$.manager_eval_notified_once') = 'true'
                                OR EXISTS (
                                        SELECT 1
                                        FROM notificaciones_app n
                                        WHERE n.tarea_id = t.id
                                            AND n.task_code = ?
                                )
                            )
    """
    cursor = conn.cursor()
    cursor.execute(sql, (task_code, employee_id, task_code))
    row = cursor.fetchone()
    cursor.close()
    return row is not None


def _mark_notified_once(conn, employee_ids: list[int], task_code: str) -> None:
    unique_ids = sorted({int(eid) for eid in employee_ids if eid is not None})
    if not unique_ids:
        return

    placeholders = ','.join(['?' for _ in unique_ids])
    sql = f"""
        UPDATE tareas_gestion_talento
        SET metadata_json = JSON_MODIFY(
            JSON_MODIFY(COALESCE(metadata_json, '{{}}'), '$.manager_eval_notified_once', 'true'),
            '$.manager_eval_notified_at',
            CONVERT(VARCHAR(19), GETDATE(), 126)
        )
        WHERE task_code = ?
          AND employee_id IN ({placeholders})
          AND status IN ('PENDIENTE', 'EN_PROCESO')
    """
    cursor = conn.cursor()
    cursor.execute(sql, [task_code, *unique_ids])
    conn.commit()
    cursor.close()


def _get_task_ids_for_employees(conn, employee_ids: list[int], task_code: str) -> list[dict]:
    if not employee_ids:
        return []
    placeholders = ','.join(['?' for _ in employee_ids])
    sql = f"""
        SELECT id, employee_id, employee_name, employee_number
        FROM tareas_gestion_talento
        WHERE task_code = ?
          AND employee_id IN ({placeholders})
          AND status IN ('PENDIENTE', 'EN_PROCESO')
    """
    cursor = conn.cursor()
    cursor.execute(sql, [task_code, *employee_ids])
    rows = cursor.fetchall()
    cursor.close()
    return [{'id': r[0], 'employee_id': r[1], 'employee_name': r[2] or '', 'employee_number': r[3] or ''}
            for r in rows]


def _build_manager_html(manager_name: str, subordinates: list, today_str: str) -> str:
    rows = ''
    for s in subordinates:
        eval_fmt = datetime.strptime(s['eval_date'], '%Y-%m-%d').strftime('%d/%m/%Y') if s.get('eval_date') else '-'
        rows += (
            f'<tr>'
            f'<td style="padding:10px 14px;border-bottom:1px solid #f3f4f6;font-size:13px;color:#374151;">{escape(s["employee_number"])}</td>'
            f'<td style="padding:10px 14px;border-bottom:1px solid #f3f4f6;font-size:13px;color:#1f1754;font-weight:600;">{escape(s["employee_name"])}</td>'
            f'<td style="padding:10px 14px;border-bottom:1px solid #f3f4f6;font-size:13px;color:#6b7280;">{escape(s["ciclo_label"])}</td>'
            f'<td style="padding:10px 14px;border-bottom:1px solid #f3f4f6;font-size:13px;color:#6b7280;">{eval_fmt}</td>'
            f'</tr>'
        )

    mgr_safe = escape(manager_name or 'Estimado/a jefe')
    return f"""<!DOCTYPE html><html lang="es"><head><meta charset="UTF-8"></head>
<body style="font-family:Segoe UI,Tahoma,Arial,sans-serif;background:#f1f0fb;padding:24px;">
<table width="100%" style="max-width:720px;margin:0 auto;background:#fff;border:1px solid #e2daff;border-radius:12px;padding:20px;">
<tr><td>
  <h2 style="margin:0;color:#1f1754;">Recordatorio: Evaluación de colaboradores a tu cargo</h2>
  <p style="color:#6b7280;margin:4px 0 16px;">Generado el {today_str}</p>
</td></tr>
<tr><td>
  <p style="color:#374151;">Hola {mgr_safe},</p>
  <p style="color:#374151;">El/los siguiente(s) colaborador(es) bajo tu supervisión tiene(n) lista su <strong>evaluación y autoevaluación</strong> de desempeño. Te adjuntamos los documentos generados para que los completes.</p>
</td></tr>
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" style="border-collapse:collapse;border:1px solid #e5e7eb;">
    <thead><tr style="background:#f9fafb;">
      <th style="padding:8px;text-align:left;">No. Empleado</th>
      <th style="padding:8px;text-align:left;">Nombre</th>
      <th style="padding:8px;text-align:left;">Tipo</th>
      <th style="padding:8px;text-align:left;">Fecha evaluación</th>
    </tr></thead>
    <tbody>{rows}</tbody>
  </table>
</td></tr>
<tr><td>
  <p style="color:#6b7280;font-size:12px;margin-top:20px;">
    Este mensaje es generado automáticamente por el sistema de Gestión de Talento.
    Si tienes dudas, contacta al área de Recursos Humanos.
  </p>
</td></tr>
</table></body></html>"""


def _d(raw):
    if isinstance(raw, datetime): return raw.date()
    if isinstance(raw, date): return raw
    try: return date.fromisoformat(str(raw)[:10])
    except Exception: return None


def run():
    log.info('==== INICIO NOTIFICACION JEFE EVALUACION ====')
    today = date.today()

    conn = get_connection()

    sql = """
    SELECT
        e.id AS employee_id,
        e.employee_number,
        e.manager_id,
        jh.first_performance_evaluation_sent AS eval1_date,
        jh.second_performance_evaluation_sent AS eval2_date,
        COALESCE(gi.full_name, '')   AS employee_name,
        COALESCE(mgi.full_name, '')  AS manager_name,
        COALESCE(mgi.corporate_email, '') AS manager_email
    FROM employees e
    INNER JOIN employee_job_history jh ON jh.employee_id = e.id
    LEFT JOIN employee_general_info gi  ON gi.employee_id  = e.id
    LEFT JOIN employee_general_info mgi ON mgi.employee_id = e.manager_id
    WHERE e.status = 'ALTA'
      AND e.manager_id IS NOT NULL
      AND (jh.first_performance_evaluation_sent IS NOT NULL
           OR jh.second_performance_evaluation_sent IS NOT NULL)
    ORDER BY mgi.full_name ASC
    """
    cursor = conn.cursor()
    cursor.execute(sql)
    rows = cursor.fetchall()
    cursor.close()

    # Agrupar por manager_email
    by_manager: dict[str, dict] = {}

    for row in rows:
        emp_id, emp_num, manager_id, eval1_raw, eval2_raw, emp_name, mgr_name, mgr_email = row
        mgr_email = (mgr_email or '').strip()
        if not mgr_email or '@' not in mgr_email:
            log.warning('Sin email válido para manager_id=%s (empleado %s) — omitido', manager_id, emp_num)
            continue

        ctx = {'eval1_raw': eval1_raw, 'eval2_raw': eval2_raw}

        for task_code, raw_key, ciclo_label, ciclo_short in CICLOS:
            eval_date = _d(ctx[raw_key])
            if not eval_date or eval_date > today:
                continue
            if (today - eval_date).days > VENTANA_NOTIFICACION_DIAS:
                log.info(
                    'Empleado %s (%s): fecha %s tiene mas de %d dias de atraso — se omite (ya gestionada por otros medios)',
                    emp_num, task_code, eval_date, VENTANA_NOTIFICACION_DIAS,
                )
                continue
            if _already_notified_once(conn, int(emp_id), task_code):
                continue

            docs = _find_evaluation_documents(conn, emp_id)
            if not docs:
                log.info('Empleado %s (%s): faltan documentos de evaluación/autoevaluación — se reintenta otro día',
                         emp_num, task_code)
                continue

            if mgr_email not in by_manager:
                by_manager[mgr_email] = {'manager_name': str(mgr_name or ''), 'items': {}, 'attachments': []}

            by_manager[mgr_email]['items'].setdefault(task_code, []).append({
                'employee_id':     emp_id,
                'employee_number': str(emp_num or ''),
                'employee_name':   str(emp_name or ''),
                'eval_date':       str(eval_date),
                'ciclo_label':     ciclo_short,
            })
            by_manager[mgr_email]['attachments'].append(docs['evaluacion'])
            by_manager[mgr_email]['attachments'].append(docs['autoevaluacion'])

    if not by_manager:
        log.info('No hay jefes con colaboradores listos para notificar. Sin envíos.')
        conn.close()
        return

    today_fmt = today.strftime('%d/%m/%Y')
    sent = 0

    for mgr_email, data in by_manager.items():
        items_by_code = data['items']
        attachments    = data['attachments']

        subordinates = [s for items in items_by_code.values() for s in items]

        # Deduplicar adjuntos
        unique_attachments, seen_paths = [], set()
        for att in attachments:
            p = att.get('path', '')
            if p and p not in seen_paths:
                seen_paths.add(p)
                unique_attachments.append(att)

        html = _build_manager_html(data['manager_name'], subordinates, today_fmt)
        plain = f"Tienes {len(subordinates)} colaborador(es) con evaluación lista para revisar. Adjuntos: {len(unique_attachments)}."
        subject = 'Recordatorio: Evaluación de tu(s) colaborador(es)'

        send_email_smtp([mgr_email], subject, html, plain, unique_attachments)
        log.info('Email enviado a %s (%s) — %d colaborador(es), adjuntos=%d',
                 mgr_email, data['manager_name'], len(subordinates), len(unique_attachments))

        for task_code, items in items_by_code.items():
            employee_ids = [s['employee_id'] for s in items]
            tasks_with_ids = _get_task_ids_for_employees(conn, employee_ids, task_code)
            if tasks_with_ids:
                insert_notificaciones(conn, tasks_with_ids, [mgr_email], task_code,
                                      message_template='Evaluación pendiente — {employee_name}')
            _mark_notified_once(conn, employee_ids, task_code)

        sent += 1

    log.info('==== FIN NOTIFICACION JEFE EVALUACION — %d emails enviados ====', sent)
    conn.close()


if __name__ == '__main__':
    run()

File: scripts/notificar_jefe_evaluacion.php
<?php

/**
 * notificar_jefe_evaluacion.php
 *
 * Envía un correo al jefe directo de cada colaborador cuya primera
 * evaluacion cae en la semana actual o proxima, recordandole que debe
 * completar la evaluacion y la autoevaluacion de su subordinado.
 *
 * Origen de datos:
 *  - employees.manager_id  → ForeignKey al empleado jefe
 *  - employee_general_info.corporate_email del jefe
 * Solo colaboradores con status = 'ALTA'.
 */

date_default_timezone_set('America/Mexico_City');

require_once __DIR__ . '/../config/sqlserver.php';
require_once __DIR__ . '/mailer.php';

function log_msg($msg) {
    $line = date('Y-m-d H:i:s') . ' | ' . $msg . PHP_EOL;
    file_put_contents(__DIR__ . '/../logs/notificacion_jefe_evaluacion.log', $line, FILE_APPEND);
}

function startOfWeek(DateTime $date): DateTime {
    $copy = clone $date;
    $copy->setTime(0, 0, 0);
    $dayOfWeek = (int)$copy->format('N');
    $copy->modify('-' . ($dayOfWeek - 1) . ' days');
    return $copy;
}

function generatedDocsRootCandidates(): array {
    $candidates = [];

    $envRoot = getenv('GENERATED_DOCS_ROOT');
    if (is_string($envRoot) && trim($envRoot) !== '') {
        $candidates[] = trim($envRoot);
    }

    $envDir = getenv('GENERATED_DOCS_DIR');
    if (is_string($envDir) && trim($envDir) !== '') {
        $candidates[] = trim($envDir);
    }

    // Ruta local tipica cuando backend y sync viven bajo c:/apache/htdocs
    $candidates[] = dirname(__DIR__, 2) . DIRECTORY_SEPARATOR . 'talent-management-app' . DIRECTORY_SEPARATOR . 'backend' . DIRECTORY_SEPARATOR . 'generated_documents';
    // Ruta por defecto dentro del contenedor backend
    $candidates[] = '/app/media/generated_documents';
    // Fallback relativo al proyecto de scripts
    $candidates[] = dirname(__DIR__) . DIRECTORY_SEPARATOR . 'generated_documents';

    $unique = [];
    $seen = [];
    foreach ($candidates as $path) {
        $normalized = rtrim((string)$path, "\\/");
        if ($normalized === '' || isset($seen[$normalized])) {
            continue;
        }
        $seen[$normalized] = true;
        $unique[] = $normalized;
    }

    return $unique;
}

function resolveGeneratedFilePath(string $relativePath): ?string {
    $relativePath = trim(str_replace(['\\', '/'], DIRECTORY_SEPARATOR, $relativePath));
    if ($relativePath === '') {
        return null;
    }

    foreach (generatedDocsRootCandidates() as $root) {
        $fullPath = $root . DIRECTORY_SEPARATOR . $relativePath;
        if (is_file($fullPath)) {
            return $fullPath;
        }
    }

    return null;
}

function findEvaluationAttachmentForEmployee($sqlsrv, int $employeeId): ?array {
    $sql = "
        SELECT TOP 1 relative_path, file_name
        FROM employee_generated_documents
        WHERE employee_id = ?
          AND (
                file_name COLLATE Latin1_General_CI_AI LIKE '%evaluacion%desempeno%'
                OR template_name COLLATE Latin1_General_CI_AI LIKE '%evaluacion%desempeno%'
          )
                ORDER BY
                        CASE WHEN LOWER(LTRIM(RTRIM(COALESCE(category, '')))) = 'alta' THEN 0 ELSE 1 END,
                        updated_at DESC,
                        created_at DESC
    ";

    $result = sqlsrv_query($sqlsrv, $sql, [$employeeId]);
    if ($result === false) {
        return null;
    }

    $row = sqlsrv_fetch_array($result, SQLSRV_FETCH_ASSOC);
    sqlsrv_free_stmt($result);

    if (!$row || empty($row['relative_path'])) {
        return null;
    }

    $fullPath = resolveGeneratedFilePath((string)$row['relative_path']);
    if (!$fullPath) {
        return null;
    }

    return [
        'path' => $fullPath,
        'name' => (string)($row['file_name'] ?? basename($fullPath)),
    ];
}

function buildManagerHtml(string $managerName, array $subordinates, string $today): string {
    $rows = '';
    foreach ($subordinates as $s) {
        $hireFmt  = !empty($s['hire_date'])  ? date('d/m/Y', strtotime((string)$s['hire_date']))  : '-';
        $evalFmt  = !empty($s['eval_date'])  ? date('d/m/Y', strtotime((string)$s['eval_date']))  : '-';
        $bucketLabel = match($s['bucket']) {
            'ACTUAL'  => '<span style="color:#d97706;font-weight:600;">Esta semana</span>',
            'PROXIMA' => '<span style="color:#2563eb;">Proxima semana</span>',
            'VENCIDA' => '<span style="color:#dc2626;font-weight:600;">Vencida</span>',
            default   => htmlspecialchars((string)$s['bucket']),
        };
        $docLabel = !empty($s['has_attachment'])
            ? '<span style="color:#166534;font-weight:600;">Adjunto</span>'
            : '<span style="color:#9ca3af;">No encontrado</span>';
        $rows .= '<tr>'
            . '<td style="padding:10px 14px;border-bottom:1px solid #f3f4f6;font-size:13px;color:#374151;">' . htmlspecialchars((string)$s['employee_number']) . '</td>'
            . '<td style="padding:10px 14px;border-bottom:1px solid #f3f4f6;font-size:13px;color:#1f1754;font-weight:600;">' . htmlspecialchars((string)$s['employee_name']) . '</td>'
            . '<td style="padding:10px 14px;border-bottom:1px solid #f3f4f6;font-size:13px;color:#6b7280;">' . $hireFmt . '</td>'
            . '<td style="padding:10px 14px;border-bottom:1px solid #f3f4f6;font-size:13px;color:#6b7280;">' . $evalFmt . '</td>'
            . '<td style="padding:10px 14px;border-bottom:1px solid #f3f4f6;font-size:13px;">' . $bucketLabel . '</td>'
            . '<td style="padding:10px 14px;border-bottom:1px solid #f3f4f6;font-size:13px;">' . $docLabel . '</td>'
            . '</tr>';
    }

    $managerNameSafe = htmlspecialchars($managerName ?: 'Estimado/a jefe');

    return '<!DOCTYPE html><html lang="es"><head><meta charset="UTF-8"></head>'
        . '<body style="font-family:Segoe UI,Tahoma,Arial,sans-serif;background:#f1f0fb;padding:24px;">'
        . '<table width="100%" style="max-width:720px;margin:0 auto;background:#fff;border:1px solid #e2daff;border-radius:12px;padding:20px;">'

        . '<tr><td>'
        . '<h2 style="margin:0;color:#1f1754;">Recordatorio: Evaluacion de colaboradores a tu cargo</h2>'
        . '<p style="color:#6b7280;margin:4px 0 16px;">Generado el ' . $today . '</p>'
        . '</td></tr>'

        . '<tr><td>'
        . '<p style="color:#374151;">Hola ' . $managerNameSafe . ',</p>'
        . '<p style="color:#374151;">El siguiente colaborador bajo tu supervision tiene su <strong>primera evaluacion y autoevaluacion</strong> proxima o en curso. '
        . 'Por favor completa tanto la <strong>evaluacion de desempeno</strong> como la <strong>autoevaluacion</strong> en las fechas indicadas.</p>'
        . '</td></tr>'

        . '<tr><td>'
        . '<table width="100%" cellpadding="0" cellspacing="0" style="border-collapse:collapse;border:1px solid #e5e7eb;">'
        . '<thead><tr style="background:#f9fafb;">'
        . '<th style="padding:8px;text-align:left;">No. Empleado</th>'
        . '<th style="padding:8px;text-align:left;">Nombre</th>'
        . '<th style="padding:8px;text-align:left;">Fecha ingreso</th>'
        . '<th style="padding:8px;text-align:left;">Fecha evaluacion (dia 21)</th>'
        . '<th style="padding:8px;text-align:left;">Urgencia</th>'
        . '<th style="padding:8px;text-align:left;">Documento evaluacion</th>'
        . '</tr></thead>'
        . '<tbody>' . $rows . '</tbody>'
        . '</table>'
        . '</td></tr>'

        . '<tr><td>'
        . '<p style="color:#6b7280;font-size:12px;margin-top:20px;">'
        . 'Este mensaje es generado automaticamente por el sistema de Gestion de Talento. '
        . 'Si tienes dudas, contacta al area de Recursos Humanos.'
        . '</p>'
        . '</td></tr>'

        . '</table></body></html>';
}

try {
    $sqlsrv = getSqlServerConnection();

    $today      = new DateTime('today');
    $weekStart  = startOfWeek($today);
    $weekEnd    = (clone $weekStart)->modify('+6 days');
    $nextStart  = (clone $weekStart)->modify('+7 days');
    $nextEnd    = (clone $weekStart)->modify('+13 days');

    $todayStr = $today->format('Y-m-d');
    $weekStartStr = $weekStart->format('Y-m-d');
    $weekEndStr   = $weekEnd->format('Y-m-d');
    $nextStartStr = $nextStart->format('Y-m-d');
    $nextEndStr   = $nextEnd->format('Y-m-d');

    log_msg("==== INICIO NOTIFICACION JEFE EVALUACION ====");
    log_msg("Ventana: ACTUAL $weekStartStr a $weekEndStr | PROXIMA $nextStartStr a $nextEndStr");

    // Traemos empleados ALTA con primera evaluacion en ventana ACTUAL o PROXIMA,
    // junto con datos de su jefe directo (manager_id).
    $sql = "
    SELECT
        e.id           AS employee_id,
        e.employee_number,
        e.manager_id,
        jh.hire_date,
        jh.first_performance_evaluation_sent AS eval_date,
        COALESCE(gi.full_name, '')            AS employee_name,
        COALESCE(mgi.full_name, '')           AS manager_name,
        COALESCE(mgi.corporate_email, '')     AS manager_email
    FROM employees e
    INNER JOIN employee_job_history jh ON jh.employee_id = e.id
    LEFT JOIN employee_general_info gi  ON gi.employee_id  = e.id
    LEFT JOIN employee_general_info mgi ON mgi.employee_id = e.manager_id
    WHERE jh.first_performance_evaluation_sent IS NOT NULL
      AND e.status   = 'ALTA'
      AND e.manager_id IS NOT NULL
      AND (
            -- ACTUAL: vence esta semana
            (jh.first_performance_evaluation_sent >= ? AND jh.first_performance_evaluation_sent <= ?)
            OR
            -- PROXIMA: vence la semana siguiente
            (jh.first_performance_evaluation_sent >= ? AND jh.first_performance_evaluation_sent <= ?)
            OR
            -- VENCIDA: ya paso pero aun no completada (incluir para que el jefe actue)
            jh.first_performance_evaluation_sent < ?
      )
    ORDER BY mgi.full_name ASC, jh.first_performance_evaluation_sent ASC
    ";

    $result = sqlsrv_query($sqlsrv, $sql, [
        $weekStartStr, $weekEndStr,
        $nextStartStr, $nextEndStr,
        $todayStr,
    ]);

    if ($result === false) {
        $errors = sqlsrv_errors(SQLSRV_ERR_ERRORS);
        throw new RuntimeException('Error en query: ' . (is_array($errors) && isset($errors[0]['message']) ? $errors[0]['message'] : 'desconocido'));
    }

    // Agrupar por manager_email
    $byManager = [];
    while ($row = sqlsrv_fetch_array($result, SQLSRV_FETCH_ASSOC)) {
        $managerEmail = trim((string)($row['manager_email'] ?? ''));
        if (!filter_var($managerEmail, FILTER_VALIDATE_EMAIL)) {
            log_msg("Sin email valido para manager_id={$row['manager_id']} (empleado {$row['employee_number']}) — omitido");
            continue;
        }

        if ($row['hire_date'] instanceof DateTimeInterface) $row['hire_date'] = $row['hire_date']->format('Y-m-d');
        if ($row['eval_date'] instanceof DateTimeInterface) $row['eval_date'] = $row['eval_date']->format('Y-m-d');

        // Calcular bucket de urgencia
        $evalDt = DateTime::createFromFormat('Y-m-d', (string)$row['eval_date']);
        if (!$evalDt) continue;
        $evalDt->setTime(0,0,0);
        $todayDt = clone $today; $todayDt->setTime(0,0,0);
        if ($evalDt < $todayDt) {
            $bucket = 'VENCIDA';
        } elseif ($evalDt >= $weekStart && $evalDt <= $weekEnd) {
            $bucket = 'ACTUAL';
        } else {
            $bucket = 'PROXIMA';
        }

        if (!isset($byManager[$managerEmail])) {
            $byManager[$managerEmail] = [
                'manager_name' => (string)($row['manager_name'] ?? ''),
                'subordinates' => [],
                'attachments' => [],
            ];
        }

        $attachment = findEvaluationAttachmentForEmployee($sqlsrv, (int)$row['employee_id']);
        $hasAttachment = $attachment !== null;
        if ($attachment !== null) {
            $byManager[$managerEmail]['attachments'][] = $attachment;
        }

        $byManager[$managerEmail]['subordinates'][] = [
            'employee_id'     => (int)$row['employee_id'],
            'employee_number' => (string)($row['employee_number'] ?? ''),
            'employee_name'   => (string)($row['employee_name'] ?? ''),
            'hire_date'       => (string)($row['hire_date'] ?? ''),
            'eval_date'       => (string)($row['eval_date'] ?? ''),
            'bucket'          => $bucket,
            'has_attachment'  => $hasAttachment,
        ];
    }
    sqlsrv_free_stmt($result);

    if (empty($byManager)) {
        log_msg('No hay jefes con colaboradores en ventana de evaluacion. Sin envios.');
        echo 'Sin notificaciones que enviar.' . PHP_EOL;
        sqlsrv_close($sqlsrv);
        exit(0);
    }

    $todayFormatted = date('d/m/Y');
    $sent = 0;

    foreach ($byManager as $managerEmail => $data) {
        $managerName  = $data['manager_name'];
        $subordinates = $data['subordinates'];
        $attachments  = $data['attachments'];

        // Deduplicar adjuntos por ruta
        $uniqueAttachments = [];
        $seenAttachmentPaths = [];
        foreach ($attachments as $attachment) {
            $path = (string)($attachment['path'] ?? '');
            if ($path === '' || isset($seenAttachmentPaths[$path])) {
                continue;
            }
            $seenAttachmentPaths[$path] = true;
            $uniqueAttachments[] = $attachment;
        }

        $html  = buildManagerHtml($managerName, $subordinates, $todayFormatted);
        $plain = 'Tienes ' . count($subordinates) . ' colaborador(es) con primera evaluacion proxima o vencida. '
            . 'Adjuntos enviados: ' . count($uniqueAttachments) . '. Revisa el correo HTML.';
        $subject = 'Recordatorio: Evaluacion de tu(s) colaborador(es) — Dia 21';

        send_email_smtp([$managerEmail], $subject, $html, $plain, $uniqueAttachments);
        log_msg("Email enviado a $managerEmail ($managerName) — "
            . count($subordinates)
            . ' colaborador(es), adjuntos='
            . count($uniqueAttachments));
        $sent++;
    }

    log_msg("==== FIN NOTIFICACION JEFE EVALUACION — $sent emails enviados ====");
    echo "Emails enviados: $sent" . PHP_EOL;

    sqlsrv_close($sqlsrv);
} catch (Throwable $e) {
    log_msg('ERROR: ' . $e->getMessage());
    http_response_code(500);
    echo 'Error: ' . $e->getMessage() . PHP_EOL;
}
